packages/testkit-backend/src/channel/testkit-protocol.js

```diff
@@ -0,0 +1,80 @@
+import readline from 'readline'
+import EventEmitter from 'events'
+
+export default class Protocol extends EventEmitter {
+  constructor (stream) {
+    super()
+    this._inRequest = false
+    this._request = ''
+    this._stream = stream
+    this._readlineInterface = null
+  }
+
+  start() {
+    if (!this._readlineInterface) {
+      this._readlineInterface = readline.createInterface(this._stream, null)
+      this._readlineInterface.on('line', this._processLine.bind(this))
+    }
+  }
+
+  stop () {
+    if (this._readlineInterface) {
+      this._readlineInterface.off('line', this._processLine.bind(this))
+      this._readlineInterface = null
+    }
+  }
+
+  // Called whenever a new line is received.
+  _processLine (line) {
+    switch (line) {
+      case '#request begin':
+        if (this._inRequest) {
+          throw new Error('Already in request')
+        }
+        this._inRequest = true
+        break
+      case '#request end':
+        if (!this._inRequest) {
+          throw new Error('End while not in request')
+        }
+        try {
+          this._emitRequest()
+        } catch (e) {
+          console.log('error', e)
+          this._emitError(e)
+        }
+        this._request = ''
+        this._inRequest = false
+        break
+      default:
+        if (!this._inRequest) {
+          this._emitError(new Error('Line while not in request'))
+        }
+        this._request += line
+        break
+    }
+  }
+
+  _emitError(e) {
+    this.emit('error', e)
+  }
+
+  serializeResponse (response) {
+    console.log('> writing response', response)
+    const responseStr = this._stringify(response)
+    return  ['#response begin', responseStr, '#response end'].join('\n') + '\n'
+  }
+
+  _emitRequest () {
+    const request = JSON.parse(this._request)
+    const { name, data } = request
+    console.log('> Got request ' + name, data)
+    this.emit('request', { name, data })
+  }
+  
+  _stringify (val) {
+    return JSON.stringify(val, (_, value) =>
+      typeof value === 'bigint' ? `${value}n` : value
```

Comment:
Technically, BigInt is not part of the TestKit protocol and this could, if there is a bug in the request handler implementation, lead to TestKit receiving invalid JSON, couldn't it? I'm not saying this is necessarily a problem, as TestKit would then crash. Just want to invite you to think about. Maybe there is another solution.